File: 2020/day08/sol.rs
use std::convert::TryInto;
use std::fs;

enum Operation {
    Jmp,
    Acc,
    Nop,
}

fn main() {
    let input = fs::read_to_string("input").expect("Can't read puzzle input file.");

    let instructions: Vec<_> = input
        .lines()
        .map(|line| {
            let (operation, argument) = line.split_once(' ').unwrap();
            let operation = match operation {
                "jmp" => Operation::Jmp,
                "acc" => Operation::Acc,
                _ => Operation::Nop,
            };
            let argument = argument.parse::<i32>().unwrap();
            (operation, argument)
        })
        .collect();

    let part1 = if let Err(result) = run_instructions(&instructions, None) {
        result
    } else {
        -1
    };

    println!("part 1: {}", part1);

    let mut part2 = 0;
    for i in 0..instructions.len() {
        let result = run_instructions(&instructions, Some(&(i as i32)));
        if let Ok(value) = result {
            part2 = value;
            break;
        }
    }

    println!("part 2: {}", part2);
}

fn run_instructions(
    instructions: &[(Operation, i32)],
    bad_instruction_offset: Option<&i32>,
) -> Result<i32, i32> {
    let mut acc = 0;
    let mut ip: i32 = 0;

    let mut executed = vec![false; instructions.len()];

    let mut swap = -1;
    if let Some(offset) = bad_instruction_offset {
        swap = *offset;
    }

    while !executed.get(ip as usize).unwrap() {
        let (operation, argument) = &instructions[ip as usize];

        match operation {
            Operation::Acc => acc += argument,
            Operation::Jmp => {
                if swap == -1 || ip != swap {
                    ip += argument - 1;
                }
            }
            _ => {
                if swap != -1 && ip == swap {
                    ip += argument - 1;
                }
            }
        }
        executed[ip as usize] = true;
        ip += 1;

        if ip >= instructions.len().try_into().unwrap() {
            return Ok(acc);
        }
    }

    Err(acc)
}

File: 2020/day08/sol.cpp
#include <fstream>
#include <iostream>
#include <vector>

/* from day02 */
std::vector<std::string> split_string(const std::string& target, char delimeter = ' ') {
    std::vector<std::string> out;
    int position = 0;
    size_t length = target.length();

    for (size_t i = 0; i < length; i++) {
        if (target[i] == delimeter) {
            out.push_back(std::string(target, position, i - position));
            position = i+1;
            continue;
        }
        if (i == length - 1) {
            out.push_back(std::string(target, position, length - position));
        }
    }

    return out;
}

/* maybe kinda useless when there are only 3 instructions */
enum Instruction {
                  JMP,
                  ACC,
                  NOP
};

Instruction get_instruction_type(const std::string& inst) {
    if (inst == "jmp") {
        return JMP;
    } else if (inst == "acc") {
        return ACC;
    }
    return NOP;
}

int run(const std::vector<std::string>& instructions, int64_t repair_index = -1) {
    int acc = 0;             /* accumulator */
    unsigned int ip = 0;     /* instruction pointer */

    bool *has_been_executed = new bool[ instructions.size() ]();

    while (ip < instructions.size()) {
        if (has_been_executed[ip]) {
            if (repair_index == -1) {
                break;
            } else {
                delete[] has_been_executed;
                return run(instructions, repair_index + 1);
                // if emulated program terminates (executes
                // instruction after the last one), it'll exit this
                // while loop and return acc
            }
        } else {
            has_been_executed[ip] = true;
        }

        std::string line = instructions[ip];
        auto inst_arg = split_string(line);
        auto inst = get_instruction_type(inst_arg[0]);
        int arg = std::stoi(inst_arg[1]);

        if (repair_index != -1 &&
            repair_index == ip) {
            if (inst == JMP) {
                inst = NOP;
            } else if (inst == NOP) {
                inst = JMP;
            }
        }

        switch (inst) {
            case JMP:
                ip += arg;
                break;
            case ACC:
                acc += arg;
                [[fallthrough]];
            case NOP:
            default:
                ip++;
        }
    }

    delete[] has_been_executed;
    return acc;
}

int main() {
    std::ifstream input("input");
    if (!input.is_open()) {
        std::cerr << "Can't read puzzle input file.\n";
        return -1;
    }

    std::vector<std::string> instructions;
    instructions.reserve(591);

    std::string line;
    while (std::getline(input, line)) {
        instructions.push_back(line);
    }

    int part1 = run(instructions);
    int part2 = run(instructions, 0);

    std::cout << "part 1: " << part1 << '\n';
    std::cout << "part 2: " << part2 << '\n';

    return 0;
}
